Generate serialization for WebCore::AppKitControlSystemImage

ddkilzer opened the pull request:
#### 9ee8898cb37eafe7d6fee49385b932dfd0ef447a
<pre>
Generate serialization for WebCore::AppKitControlSystemImage
<a href="https://bugs.webkit.org/show_bug.cgi?id=268832">https://bugs.webkit.org/show_bug.cgi?id=268832</a>
&lt;<a href="https://rdar.apple.com/122396630">rdar://122396630</a>&gt;

Reviewed by Alex Christensen.

* Source/WebCore/Modules/applepay/ApplePayLogoSystemImage.h:
* Source/WebCore/Modules/system-preview/ARKitBadgeSystemImage.h:
- Make subclasses of WebCore::SystemImage fast allocated.

* Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.h:
- Make subclass of WebCore::SystemImage fast allocated.
- Add struct IPC::ArgumentCoder as friend for encoding/decoding.

* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.h:
- Make subclass of WebCore::SystemImage fast allocated.
(WebCore::ScrollbarTrackCornerSystemImageMac::create): Add.
- Move implementation to source file.
- Declare overloaded create() method that takes instance variable
  values.
* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.mm:
(WebCore::ScrollbarTrackCornerSystemImageMac::create): Add.
- Move implementation from header to here.
- Add implementation of create() method that takes instance variable
  values.

* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
- Add support for &apos;ConstructSubclass&apos; attribute.
(SerializedType.namespace_and_name_for_construction): Add.
- Return correct class when &apos;ConstructSubclass&apos; is used.
(construct_type):
- Make use of namespace_and_name_for_construction().with
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;WebCore::AppKitControlSystemImage&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::AppKitControlSystemImage&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
- Update serialization tests.

* Source/WebKit/Shared/WebCoreArgumentCoders.h:
- Remove declarations for WebCore::AppKitControlSystemImage.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
- Define serialization for WebCore::AppKitControlSystemImage.
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm: Remove.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
- Remove WebCoreArgumentCodersMac.mm from the project.

Canonical link: <a href="https://commits.webkit.org/274354@main">https://commits.webkit.org/274354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5211fa9f52e66e5a0f94194a0258f819100c8585

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38732 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41077 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41263 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41039 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15009 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32507 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39305 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14878 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33647 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12920 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/38598 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12892 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34531 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42539 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35176 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34908 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38726 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13542 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36935 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15154 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/14673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5063 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14635 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -402,6 +402,13 @@ header: <WebCore/AppKitControlSystemImage.h>
+
+header: <WebCore/ScrollbarTrackCornerSystemImageMac.h>
+[AdditionalEncoder=StreamConnectionEncoder, RefCounted, ConstructSubclass=ScrollbarTrackCornerSystemImageMac] class WebCore::AppKitControlSystemImage {
+    [Validator='*m_controlType == WebCore::AppKitControlSystemImageType::ScrollbarTrackCorner'] WebCore::AppKitControlSystemImageType m_controlType;
```
Let's remove this from the serialization if it can only have one valid value and it must be a certain value.  Then we can remove the RELEASE_ASSERT too

ddkilzer replied:
Okay, will make that change next.  (Added `watchlist` support before I saw this.)

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -31,6 +31,7 @@
 #
 # AdditionalEncoder - generate serializers for StreamConnectionEncoder in addition to IPC::Encoder.
 # CreateUsing - use a custom function to call instead of the constructor or create.
+# ConstructSubclass - use a subclass to construct the object. Do not include namespace.
```
For any changes to the generator, add to Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in and run "make -C Source/WebKit/Scripts/webkit/tests"

aprotyas replied:
@achristensen07 it might be a good idea to document that in `generate-serializers.py`, it took me a while to figure out as well when I was making my recent change.

ddkilzer replied:
I added a `watchlist` message, and modified `metadata/git_config_extension` to show it.

Commit:
Generate serialization for WebCore::AppKitControlSystemImage
https://bugs.webkit.org/show_bug.cgi?id=268832
<rdar://122396630>

Reviewed by Alex Christensen.

* Source/WebCore/Modules/applepay/ApplePayLogoSystemImage.h:
* Source/WebCore/Modules/system-preview/ARKitBadgeSystemImage.h:
- Make subclasses of WebCore::SystemImage fast allocated.

* Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.h:
- Make subclass of WebCore::SystemImage fast allocated.
- Add struct IPC::ArgumentCoder as friend for encoding/decoding.

* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.h:
- Make subclass of WebCore::SystemImage fast allocated.
(WebCore::ScrollbarTrackCornerSystemImageMac::create): Add.
- Move implementation to source file.
- Declare overloaded create() method that takes instance variable
  values.
* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.mm:
(WebCore::ScrollbarTrackCornerSystemImageMac::create): Add.
- Move implementation from header to here.
- Add implementation of create() method that takes instance variable
  values.

* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
- Add support for 'ConstructSubclass' attribute.
(SerializedType.namespace_and_name_for_construction): Add.
- Return correct class when 'ConstructSubclass' is used.
(construct_type):
- Make use of namespace_and_name_for_construction().with
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<WebCore::AppKitControlSystemImage>::encode):
(IPC::ArgumentCoder<WebCore::AppKitControlSystemImage>::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
- Update serialization tests.

* Source/WebKit/Shared/WebCoreArgumentCoders.h:
- Remove declarations for WebCore::AppKitControlSystemImage.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
- Define serialization for WebCore::AppKitControlSystemImage.
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm: Remove.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
- Remove WebCoreArgumentCodersMac.mm from the project.

Canonical link: https://commits.webkit.org/274354@main
